### PR TITLE
Store node information in NodeInfo

### PR DESCRIPTION
**Pull request by wojtek-t**
This is significantly improving scheduler throughput.

On 1000-node cluster:
- empty cluster: ~70pods/s
- full cluster: ~45pods/s
  Drop in throughput is mostly related to priority functions, which I will be looking into next (I already have some PR #24095, but we need for more things before).

This is roughly ~40% increase.
However, we still need better understanding of predicate function, because in my opinion it should be even faster as it is now. I'm going to look into it next week.

@gmarek @hongchaodeng @xiang90

### CODE REVIEW

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 1]
```diff
@@ -104,16 +99,12 @@ func defaultPredicates() sets.String {
...
 			},
 		),
 		// Fit is determined by node selector query.
-		factory.RegisterFitPredicateFactory(
-			"MatchNodeSelector",
-			func(args factory.PluginFactoryArgs) algorithm.FitPredicate {
-				return predicates.NewSelectorMatchPredicate(args.NodeInfo)
-			},
-		),
+		// TODO: This is run as part of GeneralPredicates - figure out the way to remove it.
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: @davidopp - any idea how to solve it?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -751,27 +707,8 @@ func haveSame(a1, a2 []string) bool {
...
-	node, err := n.info.GetNodeInfo(nodeName)
-	if err != nil {
-		return false, err
-	}
-	return RunGeneralPredicates(pod, nodeName, nodeInfo, node)
-}
-
-func RunGeneralPredicates(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo, node *api.Node) (bool, error) {
-	fit, err := podFitsResourcesInternal(pod, nodeName, nodeInfo, node)
+func GeneralPredicates(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
```
Comment (hongchaodeng) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Can you add some general comments about what it does?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 3]
```diff
@@ -732,7 +686,9 @@ func getUsedPorts(pods ...*api.Pod) map[int]bool {
 	for _, pod := range pods {
 		for _, container := range pod.Spec.Containers {
 			for _, podPort := range container.Ports {
-				ports[podPort.HostPort] = true
+				if podPort.HostPort != 0 {
```
Comment (hongchaodeng) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: We should be able to aggregate this in NodeInfo.
What do you think?

File: plugin/pkg/scheduler/schedulercache/cache.go [hunk 4]
```diff
@@ -233,6 +233,44 @@ func (cache *schedulerCache) RemovePod(pod *api.Pod) error {
...
+
+	n, ok := cache.nodes[node.Name]
+	if !ok {
+		n = NewNodeInfo()
+		cache.nodes[node.Name] = n
+	}
+	return n.SetNode(node)
+}
+
+func (cache *schedulerCache) UpdateNode(oldNode, newNode *api.Node) error {
```
Comment (xiang90) [inline plugin/pkg/scheduler/schedulercache/cache.go]: what is the difference between this function and AddNode? Just naming? I would rather keep this simple for now. Adding the fast zero-allocation update path later when needed.

File: plugin/pkg/scheduler/schedulercache/node_info.go [hunk 5]
```diff
@@ -153,6 +165,16 @@ func calculateResource(pod *api.Pod) (cpu int64, mem int64, non0_cpu int64, non0
 	return
 }
 
+func (n *NodeInfo) SetNode(node *api.Node) error {
```
Comment (xiang90) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: Doc string public function?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 6]
```diff
@@ -544,14 +503,10 @@ func PodMatchesNodeLabels(pod *api.Pod, node *api.Node) bool {
...
-}
-
-func (n *NodeStatus) PodSelectorMatches(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
-	node, err := n.info.GetNodeInfo(nodeName)
-	if err != nil {
-		return false, err
+func PodSelectorMatches(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
+	node := nodeInfo.Node()
+	if node == nil {
+		return false, fmt.Errorf("node not found: %q", nodeName)
```
Comment (xiang90) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Nit: I am not sure about the k8s policy here. But it is not a good idea to return a very general but private defined error in a public function.

**Comment by hongchaodeng** [comment]
@wojtek-t 
Why is it improving the scheduler by a significant amount? Can you help explain?
From very basic point of view, it seems like GetNodeInfo() is replaced by local cache, which might reduce latency if it's remote call.

File: plugin/pkg/scheduler/factory/factory.go [hunk 7]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+			AddFunc:    c.informerAddNode,
+			UpdateFunc: c.informerUpdateNode,
+			DeleteFunc: c.informerDeleteNode,
 		},
 	)
 
 	return c
 }
 
+func (c *ConfigFactory) informerAddPod(obj interface{}) {
```
Comment (xiang90) [inline plugin/pkg/scheduler/factory/factory.go]: I feel this function name should be addPotToCache instead of informerAddPod. It is more clear about what this method really does. The informer is a consumer of this function, and when calling this function from the consumer side the consumer prefix seems to be duplicated too.

**Comment by xiang90** [comment]
@wojtek-t Thanks! The overall idea makes a lot of sense to me. Looking forward to the numbers!

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra @kubernetes/rh-scalability @smarterclayton

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 3 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Agree - just didn't get to it yet. Adding a TODO for it.

File: plugin/pkg/scheduler/schedulercache/cache.go (hunk 4 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/cache.go]: Yes - there is basically no difference now.

>
Adding the fast zero-allocation update path later when needed.

Sorry - I don't understand. Can you please clarify.

File: plugin/pkg/scheduler/schedulercache/node_info.go (hunk 5 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: done

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 6 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: We definitely shouldn't return such errors to the outside world (users). But this is more like internal scheduling function, and there are a bunch of other such examples in the code (even in this file). So I think it's ok.

File: plugin/pkg/scheduler/factory/factory.go (hunk 7 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Agree - changed.

**Comment by wojtek-t** [comment]
>
@wojtek-t 
Why is it improving the scheduler by a significant amount? Can you help explain?
From basic point of view, it seems like GetNodeInfo() is replaced by local cache, which might reduce latency if it's remote call.

@hongchaodeng - we were using local caches anyway.
What we are improving is basically lock contention.

Previously, every GetNodeInfo() call (which we were doing few times for every node in different predicate functions) was a get call to cache, which internally was taking a lock on the cache. This was just expensive.

In the few version, we do synchronizaction (locking) when updating pods/nodes on the same cache (which may potentially increase latency there, but differences are pretty small and it's not critical).
However, when scheduling, we are just acquiring lock only at the beginning to call GetNodeNameToInfoMap, and after that we don't do any locking.


BTW - my feeling is that all differences now between your benchmark and our system are a result of lock contention caused by updates coming from other components (pod & node updates).

File: plugin/pkg/scheduler/schedulercache/node_info.go [hunk 8]
```diff
@@ -153,6 +165,18 @@ func calculateResource(pod *api.Pod) (cpu int64, mem int64, non0_cpu int64, non0
...
 }
 
+// Sets the overall node information.
+func (n *NodeInfo) SetNode(node *api.Node) error {
+	n.node = node
+	return nil
+}
+
+// Removes the overall information about the node.
+func (n *NodeInfo) RemoveNode(node *api.Node) error {
```
Comment (hongchaodeng) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: Well, if we remove node,  we should just remove that entry in the cache

Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: I didn't do that on purpose - potentially there may be some race, that in the system pods are removed and then node is removed, but since pods & nodes are delivered in different watch channels, there may be a race and node deletion may be delivered first. In that case, we don't want to remove it until all pods are also removed. This is handled in cache.go

Comment (hongchaodeng) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: >  since pods & nodes are delivered in different watch channels, there may be a race and node deletion may be delivered first. 

Ah. I see. That sounds not right though. If a node is deleted, but we still know of some pods on it... Is there any explicit docs or discussion on this?

Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: I think you don't understand. If the node is deleted, there shouldn't by any pods assigned to in in db.
But we don't controller watches - they may be delivered in different orders, because those are different watches.
It's not a bug- it's like distributed system work.

Comment (hongchaodeng) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: Got that!

**Comment by wojtek-t** [comment]
Update PR description with new numbers.

**Comment by wojtek-t** [comment]
@davidopp - can you please take a look or delegate?

File: plugin/pkg/scheduler/schedulercache/cache.go (hunk 4 above)
Comment (xiang90) [inline plugin/pkg/scheduler/schedulercache/cache.go]: > Sorry - I don't understand. Can you please clarify.

I would add a update function to avoid allocation by explicitly reusing the previous allocated object. I thought this update function is here to prepare for this purpose.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 6 above)
Comment (hongchaodeng) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I would like to see a FitError (the name exists somehow in generic_scheduler) or PredicateError to encapsulate those.

Not arguing here. We can add it later to separate smaller commits.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 1 above)
Comment (hongchaodeng) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: This looks duplicate. My suggestion is to remove one?

File: plugin/pkg/scheduler/schedulercache/cache.go (hunk 4 above)
Comment (hongchaodeng) [inline plugin/pkg/scheduler/schedulercache/cache.go]: The Update seems to accomodate the controller model of Add, Update, Delete funcs.
If a node's fields was changed, e.g. status, resources, then it should be called.
Not quite useful in current use cases, I guess.

**Comment by hongchaodeng** [comment]
>  can you please take a look or delegate?

Both @xiang90 and I have reviewed it multiple times. I definitely would like to help with the review and merge.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 1 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: I'm not sure we can because of backward compatibility issues. That's why I'm asking @davidopp

File: plugin/pkg/scheduler/schedulercache/cache.go (hunk 4 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/cache.go]: We may want to optimize it at some point, but even with that it's significantly better.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 1 above)
Comment (HaiyangDING) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: I think moving this to `init()` is safe enough for backward compatibility. I should have moved it in #20204, but I forgot to do it, sorry.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 9]
```diff
@@ -732,7 +686,10 @@ func getUsedPorts(pods ...*api.Pod) map[int]bool {
 	for _, pod := range pods {
 		for _, container := range pod.Spec.Containers {
 			for _, podPort := range container.Ports {
-				ports[podPort.HostPort] = true
+				// TODO: Aggregate it at the NodeInfo level.
+				if podPort.HostPort != 0 {
```
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Why did you add this?

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 1 above)
Comment (davidopp) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: Yeah, I agree with @HaiyangDING , you can just move this to init()

File: plugin/pkg/scheduler/factory/factory.go [hunk 10]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
 		},
 	)
 
 	return c
 }
 
+func (c *ConfigFactory) addPodToCache(obj interface{}) {
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		glog.Errorf("cannot convert to *api.Pod")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: maybe print the value of obj.(type) to help debugging?

File: plugin/pkg/scheduler/factory/factory.go [hunk 11]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+	}
+	if err := c.schedulerCache.AddPod(pod); err != nil {
+		glog.Errorf("scheduler cache AddPod failed: %v", err)
+	}
+}
+
+func (c *ConfigFactory) updatePodInCache(oldObj, newObj interface{}) {
+	oldPod, ok := oldObj.(*api.Pod)
+	if !ok {
+		glog.Errorf("cannot convert to *api.Pod")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: to distinguish this from the other error, maybe say "cannot convert oldPod to *api.Pod"

File: plugin/pkg/scheduler/factory/factory.go [hunk 12]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+		return
+	}
+	newPod, ok := newObj.(*api.Pod)
+	if !ok {
+		glog.Errorf("cannot convert to *api.Pod")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: to distinguish this from the other error maybe say "cannot convert newPod to *api.Pod"

File: plugin/pkg/scheduler/factory/factory.go [hunk 13]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+func (c *ConfigFactory) deletePodFromCache(obj interface{}) {
+	var pod *api.Pod
+	switch t := obj.(type) {
+	case *api.Pod:
+		pod = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		pod, ok = t.Obj.(*api.Pod)
+		if !ok {
+			glog.Errorf("cannot convert to *api.Pod")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: maybe print the value of obj.(type) to help debugging?

File: plugin/pkg/scheduler/factory/factory.go [hunk 14]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+			return
+		}
+	default:
+		glog.Errorf("cannot convert to *api.Pod")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: maybe print the value of obj.(type) to help debugging?

File: plugin/pkg/scheduler/factory/factory.go [hunk 15]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+	}
+	if err := c.schedulerCache.RemovePod(pod); err != nil {
+		glog.Errorf("scheduler cache RemovePod failed: %v", err)
+	}
+}
+
+func (c *ConfigFactory) addNodeToCache(obj interface{}) {
+	node, ok := obj.(*api.Node)
+	if !ok {
+		glog.Errorf("cannot convert to *api.Node")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: maybe print the value of obj.(type) to help debugging?

File: plugin/pkg/scheduler/factory/factory.go [hunk 16]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+	}
+	if err := c.schedulerCache.AddNode(node); err != nil {
+		glog.Errorf("scheduler cache AddNode failed: %v", err)
+	}
+}
+
+func (c *ConfigFactory) updateNodeInCache(oldObj, newObj interface{}) {
+	oldNode, ok := oldObj.(*api.Node)
+	if !ok {
+		glog.Errorf("cannot convert to *api.Node")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: to distinguish this from the other error maybe say "cannot convert oldNode to *api.Node"

File: plugin/pkg/scheduler/factory/factory.go [hunk 17]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+		return
+	}
+	newNode, ok := newObj.(*api.Node)
+	if !ok {
+		glog.Errorf("cannot convert to *api.Node")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: to distinguish this from the other error maybe say "cannot convert newNode to *api.Node"

File: plugin/pkg/scheduler/factory/factory.go [hunk 18]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+func (c *ConfigFactory) deleteNodeFromCache(obj interface{}) {
+	var node *api.Node
+	switch t := obj.(type) {
+	case *api.Node:
+		node = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		node, ok = t.Obj.(*api.Node)
+		if !ok {
+			glog.Errorf("cannot convert to *api.Node")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: maybe print the value of obj.(type) to help debugging?

File: plugin/pkg/scheduler/factory/factory.go [hunk 19]
```diff
@@ -115,57 +117,122 @@ func NewConfigFactory(client *client.Client, schedulerName string) *ConfigFactor
...
+			return
+		}
+	default:
+		glog.Errorf("cannot convert to *api.Node")
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: maybe print the value of obj.(type) to help debugging?

File: plugin/pkg/scheduler/schedulercache/interface.go [hunk 20]
```diff
@@ -75,6 +75,15 @@ type Cache interface {
 	// RemovePod removes a pod. The pod's information would be subtracted from assigned node.
 	RemovePod(pod *api.Pod) error
 
+	// AddNode add overall information about node.
```
Comment (davidopp) [inline plugin/pkg/scheduler/schedulercache/interface.go]: s/add/adds/

File: plugin/pkg/scheduler/schedulercache/interface.go [hunk 21]
```diff
@@ -75,6 +75,15 @@ type Cache interface {
...
+	AddNode(node *api.Node) error
+
+	// UpdateNode updates overall information about node.
+	UpdateNode(oldNode, newNode *api.Node) error
+
+	// RemoveNode removes an overall information about node.
```
Comment (davidopp) [inline plugin/pkg/scheduler/schedulercache/interface.go]: s/an//

File: plugin/pkg/scheduler/schedulercache/node_info.go (hunk 8 above)
Comment (davidopp) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: Please add a comment here explaining why you do not remove the entry from the cache; I think other people may have the same question @hongchaodeng had. Also maybe mention it in cache.go

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 9 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: This gives us significant performance gains.
To clarify, currently for every container.Ports in each container, we are adding a HostPort to the "ports" result. However, if you don't have HostPort defined, it is 0 and "0" is explicitly ignored in PodFitsHostPorts which is the only function using this one.

However, since we were returning "0" in the result previously, we didn't hit an optimization of:
if len(wantPorts) == 0 {
  return true, nil
}
so we were computing ports from all pods on that node.
Which made it significantly more expensive. 

Now we simply don't do it.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 1 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 11 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 12 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 10 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Actually, you can't use "obj.(type)" outside of switch.
I will just print obj - hopefully that's good enough.

File: plugin/pkg/scheduler/factory/factory.go (hunk 13 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 14 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 15 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 16 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 17 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 18 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/factory/factory.go (hunk 19 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: done

File: plugin/pkg/scheduler/schedulercache/interface.go (hunk 20 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/interface.go]: done

File: plugin/pkg/scheduler/schedulercache/interface.go (hunk 21 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/interface.go]: done

File: plugin/pkg/scheduler/schedulercache/node_info.go (hunk 8 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/schedulercache/node_info.go]: done

**Comment by wojtek-t** [comment]
@davidopp - comments addressed.

PTAL

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 9 above)
Comment (davidopp) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I see. Can you add a comment "0 is explicitly ignored in PodFitsHostPorts, which is the only function that uses this value" ?

Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: done

**Comment by davidopp** [comment]
LGTM
